hardware_interface/test/test_resource_manager_prepare_perform_switch.cpp

```diff
@@ -0,0 +1,389 @@
+
+  // Scenarios defined by example criteria
+  std::vector<std::string> empty_keys = {};
+  std::vector<std::string> non_existing_keys = {"elbow_joint/position", "should_joint/position"};
+  std::vector<std::string> legal_keys_system = {"joint1/position", "joint2/position"};
+  std::vector<std::string> legal_keys_actuator = {"joint3/position"};
+};
+
+// System  : ACTIVE
+// Actuator: UNCONFIGURED
+TEST_F(
+  ResourceManagerPreparePerformTest,
+  when_system_active_and_actuator_unconfigured_expect_system_passing)
+{
+  preconfigure_components(
+    lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE, "active",
+    lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED, "unconfigured");
+
+  // Default behavior for empty key lists, e.g., when a Broadcaster is activated
+  EXPECT_TRUE(rm_->prepare_command_mode_switch(empty_keys, empty_keys));
+  // empty call can be done at any time and it doesn't propagates to HW components - always true
+  EXPECT_TRUE(rm_->perform_command_mode_switch(empty_keys, empty_keys));
+
+  // When TestSystemCommandModes is ACTIVE expect OK
+  EXPECT_TRUE(rm_->prepare_command_mode_switch(legal_keys_system, legal_keys_system));
+  EXPECT_EQ(claimed_system_acceleration_state_->get_value(), 1.0);
+  EXPECT_EQ(claimed_actuator_position_state_->get_value(), 0.0);
+  EXPECT_TRUE(rm_->perform_command_mode_switch(legal_keys_system, legal_keys_system));
+  EXPECT_EQ(claimed_system_acceleration_state_->get_value(), 101.0);
+  EXPECT_EQ(claimed_actuator_position_state_->get_value(), 0.0);
+
+  EXPECT_TRUE(rm_->prepare_command_mode_switch(legal_keys_system, empty_keys));
+  EXPECT_EQ(claimed_system_acceleration_state_->get_value(), 102.0);
+  EXPECT_EQ(claimed_actuator_position_state_->get_value(), 0.0);
+  EXPECT_TRUE(rm_->perform_command_mode_switch(legal_keys_system, empty_keys));
+  EXPECT_EQ(claimed_system_acceleration_state_->get_value(), 202.0);
+  EXPECT_EQ(claimed_actuator_position_state_->get_value(), 0.0);
+
+  EXPECT_TRUE(rm_->prepare_command_mode_switch(empty_keys, legal_keys_system));
+  EXPECT_EQ(claimed_system_acceleration_state_->get_value(), 203.0);
+  EXPECT_EQ(claimed_actuator_position_state_->get_value(), 0.0);
+  EXPECT_FALSE(rm_->perform_command_mode_switch(empty_keys, legal_keys_system));
+  EXPECT_EQ(claimed_system_acceleration_state_->get_value(), 303.0);
+  EXPECT_EQ(claimed_actuator_position_state_->get_value(), 0.0);
+
+  // When TestActuatorHardware is UNCONFIGURED expect OK
+  EXPECT_FALSE(rm_->prepare_command_mode_switch(legal_keys_actuator, legal_keys_actuator));
+  EXPECT_EQ(claimed_system_acceleration_state_->get_value(), 303.0);
+  EXPECT_EQ(claimed_actuator_position_state_->get_value(), 0.0);
+  EXPECT_TRUE(rm_->perform_command_mode_switch(legal_keys_actuator, legal_keys_actuator));
+  EXPECT_EQ(claimed_system_acceleration_state_->get_value(), 403.0);
+  EXPECT_EQ(claimed_actuator_position_state_->get_value(), 0.0);
```

Comment:
Does it make sense to perform the switch for already switched system?

Reply:
Yes this is normal because all hardware is always called. One has to filter interface in the hardware for now.

Reply:
> Does it make sense to perform the switch for already switched system?

Logically it doesn't make sense, but this can happen in the framework, therefore it is good to be tested.